### PR TITLE
Feature/add typescript types

### PR DESCRIPTION
**Pull request by elektronik2k5**
Fixes https://github.com/brainhubeu/react-permissible/issues/10

### CODE REVIEW

**Comment by elektronik2k5** [comment]
@kkoscielniak @piotr-s-brainhub please review :spiral_notepad:

File: package.json [hunk 1]
```diff
@@ -64,10 +65,18 @@
     "react": ">0.14.0 || >15.0.0",
     "react-dom": ">0.14.0 || >15.0.0"
   },
-  "keywords": [],
+  "keywords": [
```
Comment (kkoscielniak) [inline package.json]: Great thing, thanks!

File: types/tests/.gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+brainhubeu-react-permissible-*.tgz
+node_modules/
```
Comment (kkoscielniak) [inline types/tests/.gitignore]: Please add a new line here.

File: types/tests/Permissible.test.tsx [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import * as React from 'react'
+import { render, } from 'react-dom'
+import { Permissible, UserAndRequiredPermissions, } from '@brainhubeu/react-permissible'
+
+function callbackFunction({ userPermissions, requiredPermissions, }: UserAndRequiredPermissions){
+  console.log(`
```
[the remark this answers is not in the document]
Comment (kkoscielniak) [reply types/tests/Permissible.test.tsx]: I agree we could use more descriptive logging, but I'm against setting up `bunyan` for now, it's over-engineering for such small React lib.

File: types/tests/Permissible.test.tsx [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    requiredPermissions: ${requiredPermissions}
+  `)
+}
+
+const AccessGranted = ({ message, }: { message: string, }) => <>AccessGranted {message}</>
+
+const RestrictedComponentWithCallback = Permissible(
+  AccessGranted,
+  ['ACCESS_DASHBOARD'], // userPermissions
+  ['ACCESS_ADMIN'], // requiredPermissions
```
[the remark this answers is not in the document]
Comment (kkoscielniak) [reply types/tests/Permissible.test.tsx]: I think it's good to leave the comments, as they describe what parameters are given to the HOC. 

At some point, I want to make the params consolidated to one object with named keys. Then we will remove the comments.

File: types/tests/PermissibleRender.test.tsx [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import * as React from 'react'
+import { render, } from 'react-dom'
+import { PermissibleRender, PermissibleRenderProps, } from '@brainhubeu/react-permissible'
+
+const permissibleRenderTestProps: PermissibleRenderProps = {
+  oneperm: false,
+  children: 'restricted content',
+  userPermissions: ['view'],
```
Comment (kkoscielniak) [inline types/tests/PermissibleRender.test.tsx]: Please change the permission to `CAMEL_CASE` to make the code style similar to the one we're using already.

File: types/tests/package.json [hunk 6]
```diff
@@ -0,0 +1,27 @@
+{
```
[the remark this answers is not in the document]
Comment (kkoscielniak) [reply types/tests/package.json]: I don't know how the typescript types are generated, to be honest. 

@elektronik2k5 If it's possible to use only one `package.json`, then I think it would be way better.

File: types/tests/Permissible.test.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (kkoscielniak) [reply types/tests/Permissible.test.tsx]: @elektronik2k5 Please change the `log` to `warn` here.

File: package-lock.json [hunk 7]
```diff
@@ -0,0 +1,9797 @@
+{
```
[the remark this answers is not in the document]
Comment (elektronik2k5) [reply package-lock.json]: I also prefer `yarn`, but that's out of scope for this PR.

File: types/tests/Permissible.test.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (elektronik2k5) [reply types/tests/Permissible.test.tsx]: I don't see how it matters in a test, but sure, I'll do that.

File: types/tests/Permissible.test.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (elektronik2k5) [reply types/tests/Permissible.test.tsx]: Yeah, I follow this rule: whenever a function has more than one parameter (unless it is a binary function, such as a comparator), it should use a single object as an argument.
That's because most likely, at some point, two arguments will become three then four then all hell breaks loose :smiling_imp:.

File: types/react-permissible.d.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
+import * as React from 'react'
+
+declare module '@brainhubeu/react-permissible' {
+
```
[the remark this answers is not in the document]
Comment (elektronik2k5) [reply types/react-permissible.d.ts]: How about instead of arguing about style, I just add https://prettier.io/ and apply it incrementally to all changed files and we'll call it a day? :thinking: 

Alternative option: add prettier to the entire project, in a separate PR.

File: types/tests/package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (elektronik2k5) [reply types/tests/package.json]: In JS projects, TS types aren't generated, but are written by hand. I wrote the types in `types/react-permissible.d.ts`. This package.json file has nothing to do with writing types.

The idea of the setup is meant to mitigate a bug, which I previously solved in another [repo](https://github.com/caiogondim/fast-memoize.js/pull/69). Instead of testing types against your package source, which gets transformed before publishing, I'm testing the types against the built package using `npm pack`. This has two benefits:
1. We test against what's actually published (without publishing) - so we know it works with the real package and not just source.
2. Dependencies which are used **just for testing** the types aren't included in your package, but are scoped to the small, separate test package.

So this `types/tests/package.json` is exactly that:
1. Dependencies used only for testing TS types, which aren't even installed when a user installs the main package
2. Scripts which test TS types

I hope that clarifies it :smiley:. Let me know if you have any questions or a better suggestion on how to implement these tests.

**Comment by elektronik2k5** [comment]
@kkoscielniak I replied to most of the comments and resolved some. Please address my comments and suggestions and once we agree, I'll make the necessary code changes.

As for your comment "Also, this branch was not deployed to CI. This package uses Travis CI so it would be great to deploy it before merging." - how do I do that?

**Comment by kkoscielniak** [comment]
@elektronik2k5 

Thanks for the update! Would it be a problem if I checked it over the weekend or on beginning of next week?

File: package-lock.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (piotr-s-brainhub) [reply package-lock.json]: IMO you should just remove `package-lock.json`

**Comment by elektronik2k5** [comment]
@kkoscielniak I'm in no rush and everything can wait :). Enjoy your weekend!

File: package-lock.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (elektronik2k5) [reply package-lock.json]: Alt Text
Lock files (whether npm's `package-lock.json` or yarn's `yarn.lock`) speed up your package dependency installation time and guarantee a predictable, known-to-work dependency tree. 

Why wouldn't we want to have those advantages for free?

[the remark this answers is not in the document]
Comment (piotr-s-brainhub) [reply package-lock.json]: I don’t know about the newest NPM but a year ago I had many issues with `package-lock.json` - it was unpredictable, installing different versions for the same file and often causing a failure during installation

[the remark this answers is not in the document]
Comment (elektronik2k5) [reply package-lock.json]: Sorry to hear about your negative experience with npm :(. Yeah, versions 4 and 5 were rushed (due to competition from yarn) and had plenty of bugs. In fact, some are [still open](https://github.com/npm/npm/issues/17722)...

From my experience, the latest version (6) is more robust and works quite well. I actually recently switched the project I'm leading to yarn not because of npm (although that too), but because yarn has awesome features which npm does not.

Bottom line: I suggest we keep package-lock.json and consider moving to yarn in a separate issue/PR.

[the remark this answers is not in the document]
Comment (piotr-s-brainhub) [reply package-lock.json]: we can try that

[the remark this answers is not in the document]
Comment (kkoscielniak) [reply package-lock.json]: @elektronik2k5 We'll keep `package-lock.json` for now and consider moving to Yarn after some research if the features it provides would suit this package. 

I personally didn't have any problems with `npm` :)

File: types/tests/Permissible.test.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (kkoscielniak) [reply types/tests/Permissible.test.tsx]: Well, it's a good practice :)

File: types/tests/Permissible.test.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (kkoscielniak) [reply types/tests/Permissible.test.tsx]: Yeah, that's what I think as well, didn't have time for fixing that, though 😎

File: types/tests/package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (kkoscielniak) [reply types/tests/package.json]: That's a great explanation here. After reconsidering, I think it's not a problem to have a separate `package.json` for TS types. 

Tbh, I don't have much experience with TS types, let's say I'm the one learning here 😎

File: types/react-permissible.d.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (adam-golab) [reply types/react-permissible.d.ts]: I'd use our already configured eslint if it's possible. So that, we won't finish with the repo that half of the files are using prettier and half not. Second thing is that it's additional dependency, that we would like to avoid (I know that's just devDependency, but still something new to configure and maintain).

So, if it's possible, please follow our eslint rules, which should already report that empty line as an error.

File: types/tests/.gitignore (hunk 2 above)
Comment (elektronik2k5) [reply types/tests/.gitignore]: Done.

File: types/tests/PermissibleRender.test.tsx (hunk 5 above)
Comment (elektronik2k5) [reply types/tests/PermissibleRender.test.tsx]: Done.

**Comment by elektronik2k5** [comment]
@kkoscielniak I finally found the time to fix all the issues. Please reveiw.

**Comment by kkoscielniak** [comment]
Hi again @elektronik2k5, 

thanks for the effort (and bumping the package to React 16 as well) :)
